File: examples/platforms/android/app/src/test/java/com/here/android/test/RefEqualityTest.java
/*
 * Copyright (C) 2016-2020 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */
package com.here.android.test;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import android.os.Build;
import com.example.here.hello.BuildConfig;
import com.here.android.RobolectricApplication;
import java.util.Collections;
import java.util.List;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

@RunWith(RobolectricTestRunner.class)
@Config(
    sdk = Build.VERSION_CODES.M,
    application = RobolectricApplication.class,
    constants = BuildConfig.class)
public final class RefEqualityTest {

  @Test
  public void refEqualityPreservedForClass() {
    DummyClass instance1 = DummyFactory.getDummyClassSingleton();
    DummyClass instance2 = DummyFactory.getDummyClassSingleton();

    assertTrue(instance1 == instance2);
  }

  @Test
  public void refInequalityPreservedForClass() {
    DummyClass instance1 = DummyFactory.getDummyClassSingleton();
    DummyClass instance2 = DummyFactory.createDummyClass();

    assertFalse(instance1 == instance2);
  }

  @Test
  public void refEqualityPreservedForInterface() {
    DummyInterface instance1 = DummyFactory.getDummyInterfaceSingleton();
    DummyInterface instance2 = DummyFactory.getDummyInterfaceSingleton();

    assertTrue(instance1 == instance2);
  }

  @Test
  public void refInequalityPreservedForInterface() {
    DummyInterface instance1 = DummyFactory.getDummyInterfaceSingleton();
    DummyInterface instance2 = DummyFactory.createDummyInterface();

    assertFalse(instance1 == instance2);
  }

  @Test
  public void refEqualityPreservedForClassConstructor() {
    DummyClass instance1 = new DummyClass();
    DummyClass instance2 = DummyClass.dummyClassRoundTrip(instance1);

    assertTrue(instance1 == instance2);
  }

  @Test
  public void refInequalityPreservedForClassConstructor() {
    DummyClass instance1 = new DummyClass();
    DummyClass instance2 = new DummyClass();

    assertFalse(instance1 == instance2);
  }

  @Test
  public void refEqualityPreservedForClassInList() {
    List<DummyClass> list = Collections.singletonList(new DummyClass());

    List<DummyClass> result = DummyClass.dummyClassListRoundTrip(list);

    assertTrue(list.get(0) == result.get(0));
  }
}

File: examples/platforms/dart/test/RefEquality_test.dart
// -------------------------------------------------------------------------------------------------
// Copyright (C) 2016-2020 HERE Europe B.V.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0
// License-Filename: LICENSE
//
// -------------------------------------------------------------------------------------------------

import "package:test/test.dart";
import "package:hello/test.dart";
import "../test_suite.dart";

final _testSuite = TestSuite("RefEquality");

void main() {
  _testSuite.test("Ref equality preserved for a class", () {
    final instance1 = DummyFactory.getDummyClassSingleton();
    final instance2 = DummyFactory.getDummyClassSingleton();

    final result = instance1 == instance2;

    expect(result, isTrue);

    instance1.release();
    instance2.release();
  });
  _testSuite.test("Ref inequality preserved for a class", () {
    final instance1 = DummyFactory.getDummyClassSingleton();
    final instance2 = DummyFactory.createDummyClass();

    final result = instance1 == instance2;

    expect(result, isFalse);

    instance1.release();
    instance2.release();
  });
  _testSuite.test("Ref equality preserved for an interface", () {
    final instance1 = DummyFactory.getDummyInterfaceSingleton();
    final instance2 = DummyFactory.getDummyInterfaceSingleton();

    final result = instance1 == instance2;

    expect(result, isTrue);

    instance1.release();
    instance2.release();
  });
  _testSuite.test("Ref inequality preserved for an interface", () {
    final instance1 = DummyFactory.getDummyInterfaceSingleton();
    final instance2 = DummyFactory.createDummyInterface();

    final result = instance1 == instance2;

    expect(result, isFalse);

    instance1.release();
    instance2.release();
  });
  _testSuite.test("Ref equality preserved for a class constructor", () {
    final instance1 = DummyClass();
    final instance2 = DummyClass.dummyClassRoundTrip(instance1);

    final result = instance1 == instance2;

    expect(result, isTrue);

    instance1.release();
    instance2.release();
  });
  _testSuite.test("Ref inequality preserved for a class constructor", () {
    final instance1 = DummyClass();
    final instance2 = DummyClass();

    final result = instance1 == instance2;

    expect(result, isFalse);

    instance1.release();
    instance2.release();
  });
  _testSuite.test("Ref equality preserved for a class in a list", () {
    final list = [DummyClass()];
    final result = DummyClass.dummyClassListRoundTrip(list);

    expect(list.first == result.first, isTrue);

    list.first.release();
    result.first.release();
  });
}
